infra/build/functions/fuzzbench_local_run.py

```diff
@@ -0,0 +1,212 @@
+  if not args:
+    raise Exception(f'Error: Step {i} has no "args" field.\n')
+  if args[0] == 'push':
+    log_file.write(f'Skipping step {i} because it is a push step.\n')
+    return
+
+  step_container_work_dir = GCB_WORKSPACE_DIR + '/' + step.get('dir', '')
+
+  if args[0] == 'run' and args[1] == '-v':
+    args[2] = args[2].replace(GCB_WORKSPACE_DIR, temp_dir, 1)
+
+  docker_command = ['docker', 'run', '--rm', '--cpus=0.5']
+  docker_command.extend(['-w', step_container_work_dir])
+  docker_command.extend(['-v', f'{temp_dir}:{GCB_WORKSPACE_DIR}'])
+
+  mount_fuzzbench = any(vol.get('path') == FUZZBENCH_PATH for vol in volumes)
+  if mount_fuzzbench:
+    docker_command.extend(['-v', f'{local_fuzzbench_path}:{FUZZBENCH_PATH}'])
+
+  if image_name == DOCKER_BUILDER_IMAGE:
+    docker_command.extend(['-v', '/var/run/docker.sock:/var/run/docker.sock'])
+
+  for env_var in env_list:
+    docker_command.extend(['-e', env_var])
+
+  docker_command.append(image_name)
+  docker_command.extend(args)
+
+  # Local runs use local changes
+  if 'https://github.com/google/oss-fuzz.git' in docker_command:
+    oss_fuzz_dir = os.path.dirname(
+        os.path.dirname(
+            os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
+    docker_command = ['cp', '-r', f'{oss_fuzz_dir}', f'{temp_dir}']
+
+  if '$$' in docker_command[-1]:
+    docker_command[-1] = re.sub(r'\$\$([a-zA-Z0-9_]+)', r'"$\1"',
+                                docker_command[-1])
+
+  log_file.write(f'Executing Docker Command:\n')
+  log_file.write(
+      ' '.join(map(lambda x: f'"{x}"' if ' ' in x else x, docker_command)) +
+      '\n')
+  log_file.flush()
+
+  try:
+    start_time = time.time()
+    result = subprocess.run(docker_command,
+                            check=True,
+                            capture_output=True,
+                            text=True)
+    end_time = time.time()
+    log_file.write('--- Container STDOUT ---\n')
+    log_file.write(result.stdout)
+    log_file.write('--- Container STDERR ---\n')
+    log_file.write(result.stderr)
+    log_file.write(
+        f'--- Step {i} completed successfully --- Took {end_time - start_time}s\n\n'
+    )
+  except subprocess.CalledProcessError as e:
+    if e.returncode == 124:
+      end_time = time.time()
+      log_file.write(f'Caught timeout: {e}\n')
+      log_file.write(
+          f'--- Step {i} completed with a timeout --- Took {end_time - start_time}s\n\n'
+      )
+    else:
+      log_file.write('--- DOCKER RUN ERROR ---\n')
+      log_file.write(f'Docker command failed with exit code {e.returncode}\n')
+      log_file.write('--- Container STDOUT ---\n')
+      log_file.write(e.stdout)
+      log_file.write('--- Container STDERR ---\n')
+      log_file.write(e.stderr)
+      log_file.write(f'Failed Step Details: {step}\n')
+      log_file.write(f'Failed Docker Command: {" ".join(docker_command)}\n')
+      sys.exit(f'Execution failed at step {i}\n')
```

Comment:
I'd prefer we have one string that we format and implicitly join than so many write calls.